uplink/ecclient/client.go

```diff
@@ -473,30 +474,73 @@ func (lr *lazyPieceRanger) Size() int64 {
 // Range implements Ranger.Range to be lazily connected
 func (lr *lazyPieceRanger) Range(ctx context.Context, offset, length int64) (_ io.ReadCloser, err error) {
 	defer mon.Task()(&ctx)(&err)
-	ps, err := lr.dialPiecestore(ctx, &pb.Node{
-		Id:      lr.limit.GetLimit().StorageNodeId,
-		Address: lr.limit.GetStorageNodeAddress(),
-	})
-	if err != nil {
-		return nil, err
-	}
 
-	download, err := ps.Download(ctx, lr.limit.GetLimit(), lr.privateKey, offset, length)
-	if err != nil {
-		return nil, errs.Combine(err, ps.Close())
-	}
-	return &clientCloser{download, ps}, nil
+	reader := &lazyPieceReader{}
+	reader.lazyPieceRanger = lr
+	reader.ctx = ctx
+	reader.offset = offset
+	reader.length = length
+	return reader, nil
 }
 
-type clientCloser struct {
+type lazyPieceReader struct {
+	*lazyPieceRanger
+	ctx    context.Context
+	offset int64
+	length int64
+
+	mu sync.RWMutex
+
+	dialed bool
+	closed bool
 	piecestore.Downloader
 	client *piecestore.Client
 }
 
-func (client *clientCloser) Close() error {
+func (lr *lazyPieceReader) Read(data []byte) (_ int, err error) {
+	lr.mu.RLock()
+	defer lr.mu.RUnlock()
+
+	if lr.closed {
+		return 0, io.EOF
+	}
+	if !lr.dialed {
+		mon.Task()(&lr.ctx)(&err)
```

Comment:
How about extracting the dialing logic in a separate method?

Reply:
Also defer is missing.